### PR TITLE
feat: Determine ordering of file groups

### LINKED ISSUE #7490: Use file statistics in query planning to avoid sorting when unecessary
### Is your feature request related to a problem or challenge?

Related issue: #6672

DataFusion currently cannot avoid sorts when there are more files than there are `target_partitions`. 

When querying data from an object store, DataFusion will attempt to group files into file groups internally, which are executed concurrently with one another. However, if any file group contains multiple files, the `file_sort_order` of the files cannot be maintained at present. 

To illustrate, suppose we had a table `trades` comprised of the following files in S3:

```sh
❯ mc ls s3/trades/year=2023/month=01 -r
[2023-05-23 15:13:18 CDT] 1.2GiB STANDARD day=03/trades-2023-01-03.parquet
[2023-05-23 15:12:36 CDT] 1.2GiB STANDARD day=04/trades-2023-01-04.parquet
[2023-05-23 15:12:00 CDT] 1.1GiB STANDARD day=05/trades-2023-01-05.parquet
[2023-05-23 15:13:21 CDT] 1.2GiB STANDARD day=06/trades-2023-01-06.parquet
[2023-05-23 15:12:40 CDT] 1.2GiB STANDARD day=09/trades-2023-01-09.parquet
```

and their sort order was `timestamp ASC`. If our `target_partitions` is set to `3`, the following query:
```sql
SELECT * 
FROM trades 
ORDER BY timestamp ASC
LIMIT 50000;
```

results in the following suboptimal plan:

```
GlobalLimitExec: skip=0, fetch=50000
  SortPreservingMergeExec: [timestamp@0 ASC NULLS LAST], fetch=50000
    SortExec: fetch=50000, expr=[timestamp@0 ASC NULLS LAST]
      ParquetExec: file_groups={3 groups: [[year=2023/month=01/day=05/trades-2023-01-05.parquet, year=2023/month=01/day=06/trades-2023-01-06.parquet], [year=2023/month=01/day=03/trades-2023-01-03.parquet, year=2023/month=01/day=09/trades-2023-01-09.parquet], [year=2023/month=01/day=04/trades-2023-01-04.parquet]]}, projection=[ticker, timestamp, participant_timestamp, trf_timestamp, sequence_number, conditions, id, price, size, correction, exchange, trf, tape, year, month, day]
```

DataFusion has decided that this plan needs a sort, because some file groups have multiple files. (This could be avoided if there was only 1 or even up to 3 files.) However, in this case we know that `trades-2023-01-03.parquet` could be streamed before `trades-2023-01-04.parquet` in timestamp order, because every timestamp in `trades-2023-01-03.parquet` precedes every timestamp in `trades-2023-01-04.parquet`. In fact, the physical plan shown above *is* ordered -- but DataFusion does not know this and currently has no way *to* know this. 

### Describe the solution you'd like

Essentially, DataFusion should be able to detect which files are non-overlapping, and use this to intelligently distribute files into file groups in such a way that still outputs data in order. Below I offer one possible path to doing so, which I believe should be minimally invasive. 

At a minimum, [`PartitionedFile`](https://docs.rs/datafusion/latest/datafusion/datasource/listing/struct.PartitionedFile.html) should have an additional *optional* field, `statistics`, which contains a [`Statistics`](https://docs.rs/datafusion/latest/datafusion/common/struct.Statistics.html) object with the min/max statistics for that file. [`FileScanConfig::project`](https://docs.rs/datafusion/latest/datafusion/datasource/physical_plan/struct.FileScanConfig.html#method.project) should be changed to detect when files within a file group are distributed in order. Lastly, a physical optimizer to redistribute file groups to be ordered may be necessary to take advantage of this in some cases. 

This does not solve the issue of how to feed file-level statistics into DataFusion, but users may add extensions to DataFusion that do so -- for example a custom [`TableProvider`](https://docs.rs/datafusion/latest/datafusion/datasource/provider/trait.TableProvider.html) could do this. However, it should be feasible to integrate this feature into [`ListingTable`](https://docs.rs/datafusion/latest/datafusion/datasource/listing/struct.ListingTable.html). In fact, when `collect_statistics` is enabled, the `ListingTable` already fetches file-level statistics on each query, but discards them after rolling them up into one statistic per column. 

### Describe alternatives you've considered

At my company, we created a custom `FileFormat` implementation that outputs a wrapped `ParquetExec` with the `output_ordering()` method overrided, and the files redistributed to be in-order. However, instead of using statistics, it relies on hints from configuration provided by the user, plus this does not particularly seem in the spirit of what `FileFormat` is supposed to be. We would like to implement this optimization in a way that fits better with DataFusion and works out of the box without hints. 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by suremarc**
## Which issue does this PR close?

Closes #7490 .

## Rationale for this change

See details in #7490 - this feature helps DataFusion eliminate sorts when files can be shown to be non-overlapping in terms of min/max statistics.

## What changes are included in this PR?

* Add a new `FileScanConfig::sort_file_groups` method that distribute files via a bin packing algorithm, ensuring that no two files have overlapping statistics
* Make `FileScanConfig::project` check if file groups are sorted when determining projected output orderings
* Add a new internal `MinMaxStatistics` struct that uses the Arrow Row API to efficiently sort & compare file statistics.

## Are these changes tested?

Yes - there is a unit test and a sqllogictest. 

## Are there any user-facing changes?

Yes - there is a new optional `statistics` field in `PartitionedFile`, which is part of the proposal in #7490. 

There is also the new `FileScanConfig::sort_file_groups` API

### CODE REVIEW

**Comment by Dandandan** [comment]
@suremarc sorry for the noise, just trying to run the benchmark command!

**Comment by Dandandan** [comment]
Running /benchmark once more to see if previous run was noise

**Comment by gruuya** [comment]
Fwiw, I've noticed that it seem to have a minor systematic bias against the PR results in general (even if e.g. there are literally no code changes in it). This manifests like above, by reporting a couple of queries being marginally slower.

I think this will be alleviated once a dedicated runner is used, and/or when more benchmark types are included.

**Comment by Dandandan** [comment]
Yeah seems noise, different queries running slower in second run.

**Review by alamb — COMMENTED** [review]
Thank you @suremarc  -- this looks very cool


@NGA-TRAN  can you help review this PR (we have a version of this code upstream in InfludDB, as well as a specialized "ProgressiveEval" operator that takes advantage of non-overlapping column ranges

**Comment by NGA-TRAN** [comment]
I will review this either today or tomorrow

**Comment by suremarc** [comment]
@alamb @NGA-TRAN 

Thanks for the speedy feedback - I guess I'll go ahead and mark this as ready for review since it's being reviewed, haha.  

I'm still planning to add an end-to-end test with `ListingTable` but just haven't figured out how to do it yet

**Comment by NGA-TRAN** [comment]
Sorry I could not get to this today. I will try to review it tomorrow/later this week

**Comment by NGA-TRAN** [comment]
I am reviewing this now

**Comment by suremarc** [comment]
@NGA-TRAN I added a sqllogictest and fixed some bugs in the implementation. CI [seems to be passing now](https://github.com/polygon-io/arrow-datafusion/actions/runs/8370873137)

**Comment by NGA-TRAN** [comment]
Thanks @suremarc . I am still in the middle of the review. Sorry it takes more time. So far the code looks really good

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (NGA-TRAN) [inline datafusion/core/src/datasource/physical_plan/file_scan_config.rs:209]: I understand the purpose of this is to split and group files.  Files in each group is sorted and not overlapped. If there are many groups, each of them must overlap with at least one other.

Is this the first step for following PRs to do something with these sorted file groups? Like files in each group will be read sequentially without going thru  SortPreservingMerge; and then data streams of different groups will go thru SortPreservingMerge?

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (NGA-TRAN) [inline datafusion/core/src/datasource/physical_plan/file_scan_config.rs:888]: Can we add more tests here?
1. The same input but "2" is now in the middle to ensure we sort and group them correctly
2. All three non-overlapped files
3. All three overlapped files
4. Empty input

**Review by NGA-TRAN — APPROVED** [review]
The code looks good. I just ask for a few more unit tests and what is the next step.

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (suremarc) [reply datafusion/core/src/datasource/physical_plan/file_scan_config.rs:209]: If I understand correctly what you're asking, I think it will already do this -- see for example the sqllogictest I added: https://github.com/apache/arrow-datafusion/blob/e9fad54257a13ab9630a0f038fa2b5a1b26aa006/datafusion/sqllogictest/test_files/parquet.slt#L193-L203

The `FileScanConfig` correctly detects that the files are ordered based on their statistics, and this information is propagated to the physical planner. So fortunately this works already. 

As for next steps, it actually just occurred to me that if there are 1000 files that all overlap, after this PR `ListingTable` will produce 1000 file groups which is really bad. So I think I should add a check that only uses the sorted file groups if they produce less than `target_partitions` file groups. I will add that in this PR. 

The other thing we should maybe do is modify `sort_file_groups` to _increase_ the number of file groups if it's lower than `target_partitions`, but I also thought maybe we should let DataFusion's `EnforceDistribution` physical optimizer take care of this, so it might not be necessary. 

Other than that I'm not sure immediately sure what next steps are, without gathering more data from real-world usage. Though, I would like to add some benchmarks to DataFusion for large plans (with thousands of files), as I think the performance of this code might become more apparent

Comment (NGA-TRAN) [reply datafusion/core/src/datasource/physical_plan/file_scan_config.rs:209]: Thanks. I think another possible next step is to move the implementation of `ProgressiveEvalExec` from InfluxDB IOx here and then  someone can use it and the results of this PR to do https://github.com/apache/arrow-datafusion/issues/6672. 
What do you think @alamb ?

File: datafusion/core/src/datasource/listing/mod.rs
Comment (alamb) [inline datafusion/core/src/datasource/listing/mod.rs:70]: I think we should also provide some more guidance here for example

```suggestion
    /// Optional statistics that describe the data in this file if known.
    ///
    /// DataFusion relies on these statistics for planning so if they are incorrect
    /// incorrect answers may result.
```

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_scan_config.rs:145]: I wonder if we can avoid creating the entire schema here and instead reuse it from elsewhere -- the information should already be present to plan queries I think. As we try to make planning more efficient avoiding re-creating the same schema over and over again I think will be importatn

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_scan_config.rs:210]: maybe we could name this function something more related to the non overlapping nature

Perhaps like `split_groups_by_statistics` or something?

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_scan_config.rs:234]: Having a bit more context would be helpful I think

```suggestion
        .map_err(|e| e.context("construct min/max statistics for sort_file_groups"))?;
```

File: datafusion/core/src/datasource/physical_plan/mod.rs
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/mod.rs:548]: I don't think we can just use `NULL` when the statistics are not present (as NULL will sort somewhere in the sort order, but in reality the data may be anywhere / possibly overlap) -- Thus I think this code needs to return an error in this case

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (alamb) [reply datafusion/core/src/datasource/physical_plan/file_scan_config.rs:888]: Other important cases are additional negative cases
1. The  files have values but no statistics (the groups shouldn't get reordered)
2. One of the files has no statistics

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_scan_config.rs:866]: why is `allow(unused)` needed?

File: datafusion/core/src/datasource/physical_plan/mod.rs
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/mod.rs:462]: Since creating `MinMaxStatistics` is likely relatively expensive I wonder if there is some way to avoid it being recreated twice

File: datafusion/core/src/datasource/physical_plan/mod.rs
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/mod.rs:511]: I think it would help to explain the requirements on this structure for correctness:
1. That all rows in the sort order have statistics for all values

File: datafusion/sqllogictest/test_files/parquet.slt
Comment (alamb) [inline datafusion/sqllogictest/test_files/parquet.slt:178]: I would like to request moving the tests for this optimization to its own .slt file (e.g. `sorted_parquet.slt` or something). It is important enough and tricky enough I think to warrant specialized tests

File: datafusion/sqllogictest/test_files/parquet.slt
Comment (alamb) [inline datafusion/sqllogictest/test_files/parquet.slt:194]: I don't fully understand this comment -- this is a new test added as part of that PR, right?

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (alamb) [reply datafusion/core/src/datasource/physical_plan/file_scan_config.rs:209]: > As for next steps, it actually just occurred to me that if there are 1000 files that all overlap, after this PR ListingTable will produce 1000 file groups which is really bad. So I think I should add a check that only uses the sorted file groups if they produce less than target_partitions file groups. I will add that in this PR.

I agree


> Thanks. I think another possible next step is to move the implementation of ProgressiveEvalExec from InfluxDB IOx here and then someone can use it and the results of this PR to do #6672.

I think it is a good idea @NGA-TRAN  -- @suremarc and anyone else following along `ProgressiveEvalExec` is an operator we implemented in InfluxDB 3.0 to replace `SortPreservingMerge` in plans where we know the inputs don't overlap in sort-key space (what this PR calculates) and that avoids the merge as well as reads the files more sequentially

> Other than that I'm not sure immediately sure what next steps are, without gathering more data from real-world usage. Though, I would like to add some benchmarks to DataFusion for large plans (with thousands of files), as I think the performance of this code might become more apparent

I agree a real world benchmark would be great. As the "running queries on sorted parquet files" is also quite important to InfluxDB 3.0 we can probably help you with that.

Do you have data in mind? If not, we could potentially make something using https://github.com/apache/arrow-datafusion/blob/main/test-utils/src/data_gen.rs#L251 (that sorts on the first few low cardinality columns, for example)

**Review by alamb — COMMENTED** [review]
Thank you @suremarc  -- this is very cool and definitely a great feature to add. 

Thank you also @NGA-TRAN  for the review

From my perspective, the only thing this PR really needs is some additional tests (especially around missing statistics). Since it is effectively "opt in" (the code won't be activated unless a user explicitly supplies `PartitionedFile::Statistics`) I think it would be ok to merge in after that and then we can file follow on tickets for things like benchmarks, ProgressiveEval, etc

What do you think?

**Comment by suremarc** [comment]
@alamb I won't be able to get to this until tomorrow, just letting you know

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (suremarc) [reply datafusion/core/src/datasource/physical_plan/file_scan_config.rs:209]: @alamb I didn't have any data in mind, so any headstart we can get sounds great.

Regarding the `SortPreservingMerge`, I want to state again that `sort_file_groups` will produce (if possible) only one file group with all files sequentially ordered, so I think the plan will not have a sort-preserving-merge in the first place. I think maybe I just need to make this clearer in a separate sqllogictest. But crucially this means that #6672 would already be fixed by this PR.

**Comment by alamb** [comment]
> @alamb I won't be able to get to this until tomorrow, just letting you know

No worries - thanks for doing it @suremarc  -- I am traveling this week so I won't have time to review until later this week. 

Thank you so much

**Comment by alamb** [comment]
I think this PR is quite nice and it would be great if we could get the tests written and the code polished up. 

Marking as draft as I think this PR is no longer waiting on feedback. Please mark it as ready for review when it is ready for another look

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (suremarc) [reply datafusion/core/src/datasource/physical_plan/file_scan_config.rs:866]: Definitely not needed, this just got left in from earlier

File: datafusion/sqllogictest/test_files/parquet.slt
Comment (suremarc) [reply datafusion/sqllogictest/test_files/parquet.slt:194]: Yeah I guess the wording is confusing since it implies the test existed before this PR. Either way I'm going to make a new sqllogictest so I'll rewrite this comment

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (suremarc) [reply datafusion/core/src/datasource/physical_plan/file_scan_config.rs]: I think that is a good name, I'll go with that

**Review by suremarc — COMMENTED** [review]
@alamb @NGA-TRAN 

Sorry for the long delay. I've been busy the last month and there ended up being some bugs with the implementation, particularly around descending sort & projecting physical expressions. 

I've added a sqllogictest with descending sort, projections, sorting on partition columns, and overlapping files. I think it covers the edge cases pretty well. Please feel free to take a look when you have time

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (suremarc) [reply datafusion/core/src/datasource/physical_plan/file_scan_config.rs]: I managed to get rid of creating this schema, fortunately -- we can use projection (the `Vec<usize>`) to recover the positions of the statistics in the original schema.

**Comment by alamb** [comment]
@NGA-TRAN  do you have time to review this PR as well?

**Comment by alamb** [comment]
No worries @suremarc  -- I am very excited about this PR. I plan to review it sometime this week (hopefully later today)

**Comment by NGA-TRAN** [comment]
I will review this either today or tomorrow

File: datafusion/sqllogictest/test_files/parquet_sorted_statistics.slt
Comment (NGA-TRAN) [inline datafusion/sqllogictest/test_files/parquet_sorted_statistics.slt]: I am a bit surprised this query always return the first 3 rows and its dat tis sorted. Maybe you want to make it deterministic in case this behavior no longer holds in the future by using `SELECT * FROM src_table where int_col <=3 order by int_col`

I think since you make your columns have corresponding increasing data, the order by can be on any column and your data is always sorted on any column

File: datafusion/sqllogictest/test_files/parquet_sorted_statistics.slt
Comment (NGA-TRAN) [inline datafusion/sqllogictest/test_files/parquet_sorted_statistics.slt]: Similar as above,  `SELECT * FROM src_table where int_col >=4 and int_col <=6 order by int_col`

File: datafusion/sqllogictest/test_files/parquet_sorted_statistics.slt
Comment (NGA-TRAN) [inline datafusion/sqllogictest/test_files/parquet_sorted_statistics.slt]: Same as above, you want filter that provide deterministic result and data is sorted

File: datafusion/sqllogictest/test_files/parquet.slt
Comment (NGA-TRAN) [inline datafusion/sqllogictest/test_files/parquet.slt]: Nice negative test

**Review by NGA-TRAN — APPROVED** [review]
Thanks @suremarc for adding tests and all refactoring works. The PR looks good to me

File: datafusion/sqllogictest/test_files/parquet_sorted_statistics.slt
Comment (suremarc) [reply datafusion/sqllogictest/test_files/parquet_sorted_statistics.slt]: I'm not sure what happened here but I think you were looking at an old copy of the code -- GH says this is outdated. The latest version has a sort in this command

Comment (NGA-TRAN) [reply datafusion/sqllogictest/test_files/parquet_sorted_statistics.slt]: Right. See my comment here https://github.com/apache/datafusion/pull/9593#pullrequestreview-2032171364

Comment (suremarc) [reply datafusion/sqllogictest/test_files/parquet_sorted_statistics.slt]: Ah, ok that makes sense

**Comment by suremarc** [comment]
I'm not sure why changing a comment caused the tests to start failing.... oof.

**Comment by alamb** [comment]
Added API change label as I it adds a new field to `PartitionedFile`

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_scan_config.rs:234]: I wonder if we could move this into `statistics itself`  somehing like

```rust
        let indices_sorted_by_min = statistics.indices_sorted_by_min()
```

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_scan_config.rs:975]: can we please add a test for a single input file too?

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_scan_config.rs:953]: I think all these tests also always have the first file with the minimum stastistics value -- can you possibly also test what happens when it is not (aka add a test that runs this test with file ids 2, 1, 0)?

**Review by alamb — COMMENTED** [review]
Thank you so much @suremarc  -- this PR is pretty epic

I am a little worried about the lack of coverage in `MinMaxStatistics`, especially around some of the tricky edge cases with projected schemas and projections.  However this PR is already quite large and been open for quite a while

Here is how I suggest we proceed:
1.  Add a config value for this feature and default it to `false` (not enabled by default)
2. Merge this PR
3. File additional tickets / tests before enabling it
4. Add additional test coverage as follow on PRs and do some more testing.

Once we feel good about that we can make a PR to turn it on by default.

Thank you very much @NGA-TRAN  for your reviews

File: datafusion/core/src/datasource/physical_plan/mod.rs
Comment (suremarc) [reply datafusion/core/src/datasource/physical_plan/mod.rs]: I am not sure what you mean, but `projection` is what was used to produce `projected_schema`. It tells us what position the columns of `projected_schema` would be in the full schema. Does that make it more clear?

**Comment by suremarc** [comment]
@alamb I added a config value, and I moved `MinMaxStatistics` to its own module as requested. I wasn't sure if I should delay addressing your feedback on tests to the next PR, since it seems like the suggested plan is to merge this PR first.

**Comment by alamb** [comment]
> @alamb I added a config value, and I moved MinMaxStatistics to its own module as requested. I wasn't sure if I should delay addressing your feedback on tests to the next PR, since it seems like the suggested plan is to merge this PR first.

Sorry -- sounds good. I am going to give this PR another look and file some follow on tickets.

**Review by alamb — APPROVED** [review]
Thank you @suremarc  -- epic work. Let's merge this one in and keep iterating on main

Thanks again for sticking with this -- this is very exciting

File: datafusion/core/src/datasource/physical_plan/mod.rs
Comment (alamb) [reply datafusion/core/src/datasource/physical_plan/mod.rs]: I guess I was thinking of subtle bugs related to when:
1. The schema of the files is different but compatible (e.g. one file as (`time`, `date`, `symbol`) but the other file had (`date`, `symbol`, `time`) for example
2. The query orders by a subset of the columns (e.g. `ORDER BY time`)
3. The query orders by a subset of the columns that is not the sort order (`ORDER BY date`)

**Comment by alamb** [comment]
Filed https://github.com/apache/datafusion/issues/10336 to track enable this flag by default

File: datafusion/core/src/datasource/physical_plan/mod.rs
Comment (suremarc) [reply datafusion/core/src/datasource/physical_plan/mod.rs]: Oh... I didn't even think about option 1. But I was assuming that the layout of the file statistics should match the table schema and not the individual file's schema. It seems that that's what DataFusion does currently.

**Pull request merged**
